### ISSUE TITLE
Authentication failure in RESP3

### ISSUE THREAD

**Issue by yrashk** [aembke/fred.rs#54]
When trying to connect to a password-protected Redis instance using RESP3, I get the following error:

```
NOAUTH HELLO must be called with the client already authenticated, otherwise the HELLO AUTH <user> <pass> option can be used to authenticate the client and select the RESP protocol version at the same time
```

It works just fine if I turn off RESP3 and go back to RESP2.

I've devised a quick patch that seems to fix this problem by ensuring the client is authenticated before attempting to switch protocols:

```diff
diff --git a/src/protocol/connection.rs b/src/protocol/connection.rs
index f936197..5565039 100644
--- a/src/protocol/connection.rs
+++ b/src/protocol/connection.rs
@@ -332,8 +332,9 @@ pub async fn create_authenticated_connection_tls(
   let socket = TcpStream::connect(addr).await?;
   let tls_stream = tls::create_tls_connector(&inner.config)?;
   let socket = tls_stream.connect(domain, socket).await?;
-  let framed = switch_protocols(inner, Framed::new(socket, codec)).await?;
+  let framed = Framed::new(socket, codec);
   let framed = authenticate(framed, &client_name, username, password, inner.is_resp3()).await?;
+  let framed = switch_protocols(inner, framed).await?;
   let framed = select_database(inner, framed).await?;
 
   client_utils::set_client_state(&inner.state, ClientState::Connected);
@@ -360,8 +361,9 @@ pub async fn create_authenticated_connection(
   let username = inner.config.read().username.clone();
 
   let socket = TcpStream::connect(addr).await?;
-  let framed = switch_protocols(inner, Framed::new(socket, codec)).await?;
+  let framed = Framed::new(socket, codec);
   let framed = authenticate(framed, &client_name, username, password, inner.is_resp3()).await?;
+  let framed = switch_protocols(inner, framed).await?;
   let framed = select_database(inner, framed).await?;
 
   client_utils::set_client_state(&inner.state, ClientState::Connected);
```

After applying this patch, it works as intended. If this is deemed to be an appropriate solution, I am happy to send a PR, or you can apply it yourself if it is more convenient.

**Comment by aembke** [comment]
Thanks for finding this @yrashk. I have a similar change on one of my working branches for the next major version, but it's diverged too much for me to just merge that. Feel free to make a PR, otherwise I can make one this weekend. 

I'll most likely take this change, https://github.com/aembke/fred.rs/pull/53, and https://github.com/aembke/fred.rs/pull/51 and release them as 5.1.1 this weekend.

**Comment by yrashk** [comment]
@aembke I sent PR #55 with my patch. Hoping to see 5.1.1 soon!

**Comment by aembke** [comment]
This is merged and released in 5.2.0. Thanks @yrashk.
